soc/arm/nordic_nrf/nrf53/sync_rtc.c

```diff
@@ -0,0 +1,244 @@
+		/* APP: Synchronized completed */
+		free_resources(channels);
+	} else {
+		/* Compare event generated, reconfigure PPI and wait for
+		 * IPC event from APP.
+		 */
+		ppi_ipc_to_rtc(channels, true);
+	}
+}
+
+static log_timestamp_t sync_rtc_timestamp_get(void)
+{
+	return (log_timestamp_t)(sys_clock_tick_get() + nrf53_sync_offset);
+}
+
+static void ipm_callback(const struct device *ipmdev, void *user_data,
+			 uint32_t id, volatile void *data)
+{
+	extern const struct log_link *log_link_ipc_get_link(void);
+
+	union rtc_sync_channels channels;
+	int err;
+	uint32_t cc;
+
+	channels.raw = (uint32_t)user_data;
+	err = ipm_set_enabled(ipmdev, false);
+	__ASSERT_NO_MSG(err == 0);
+
+	cc = z_nrf_rtc_timer_compare_read(channels.ch.rtc);
+
+	/* Clear previous task,event */
+	ppi_ipc_to_rtc(channels, false);
+
+	if (IS_ENABLED(CONFIG_SOC_NRF5340_CPUAPP)) {
+		/* Setup new connection from RTC to IPC and set RTC to a new
+		 * interval that contains captured offset.
+		 */
+		ppi_rtc_to_ipc(channels, true);
+
+		z_nrf_rtc_timer_compare_set(channels.ch.rtc, cc + cc + RTC_SYNC_ARBITRARY_OFFSET,
+					    rtc_cb, (void *)channels.raw);
+	} else {
+		/* Synchronization completed */
+		free_resources(channels);
+		nrf53_sync_offset = cc - RTC_SYNC_ARBITRARY_OFFSET - 2 * sync_cc;
+		if (IS_ENABLED(CONFIG_NRF53_SYNC_RTC_LOG_TIMESTAMP)) {
+			uint32_t offset_us =
+				(uint64_t)nrf53_sync_offset * 1000000 /
+				sys_clock_hw_cycles_per_sec();
+
+			log_set_timestamp_func(sync_rtc_timestamp_get,
+					       sys_clock_hw_cycles_per_sec());
+			LOG_INF("Updated timestamp to synchronized RTC by %d ticks (%dus)",
+					nrf53_sync_offset, offset_us);
+		}
+	}
+}
+
+/* Setup RTC synchronization. */
+static int sync_rtc_setup(const struct device *unused)
+{
+	ARG_UNUSED(unused);
+
+	nrfx_err_t err;
+	union rtc_sync_channels channels;
+	int32_t sync_rtc_ch;
+	const struct device *ipm_dev;
+	int rv;
+
+	ipm_dev = device_get_binding("IPM_" STRINGIFY(CONFIG_NRF53_SYNC_RTC_IPM_IN));
+	if (ipm_dev == NULL) {
+		rv = -EIO;
+		goto bail;
+	}
+
+	err = nrfx_dppi_channel_alloc(&channels.ch.ppi);
+	if (err != NRFX_SUCCESS) {
+		rv = -ENODEV;
+		goto bail;
+	}
+
+	sync_rtc_ch = z_nrf_rtc_timer_chan_alloc();
+	if (sync_rtc_ch < 0) {
+		nrfx_dppi_channel_free(channels.ch.ppi);
+		rv = sync_rtc_ch;
+		goto bail;
+	}
+
+	channels.ch.rtc = (uint8_t)sync_rtc_ch;
+	channels.ch.ipc_out = CONFIG_NRF53_SYNC_RTC_IPM_OUT;
+	channels.ch.ipc_in = CONFIG_NRF53_SYNC_RTC_IPM_IN;
+
+	ipm_register_callback(ipm_dev, ipm_callback, (void *)channels.raw);
+	rv = ipm_set_enabled(ipm_dev, true);
+	__ASSERT_NO_MSG(rv == 0);
+
+	nrfx_gppi_channels_enable(BIT(channels.ch.ppi));
+
+	if (IS_ENABLED(CONFIG_SOC_NRF5340_CPUAPP)) {
+		ppi_ipc_to_rtc(channels, true);
+	} else {
+		ppi_rtc_to_ipc(channels, true);
+		sync_cc = z_nrf_rtc_timer_read() + RTC_SYNC_ARBITRARY_OFFSET;
+		z_nrf_rtc_timer_compare_set(channels.ch.rtc, sync_cc,
+					    rtc_cb, (void *)channels.raw);
```

Comment:
The use of `RTC_SYNC_ARBITRARY_OFFSET` here is a bit confusing, as the value used here has nothing to do with offset between APP and NET, it's only about successful setting of the RTC compare in the near future, right? And I think you should prevent preemption of this part of code.